AK/StringView.cpp

```diff
@@ -68,44 +68,74 @@ Vector<StringView> StringView::split_view(StringView separator, SplitBehavior sp
     return parts;
 }
 
-Vector<StringView> StringView::lines(bool consider_cr) const
+template<typename Callback>
+static void for_each_line(StringView string, Callback&& callback)
 {
-    if (is_empty())
-        return {};
-
-    if (!consider_cr)
-        return split_view('\n', SplitBehavior::KeepEmpty);
+    char const* characters = string.characters_without_null_termination();
 
-    Vector<StringView> v;
     size_t substart = 0;
     bool last_ch_was_cr = false;
-    bool split_view = false;
-    for (size_t i = 0; i < length(); ++i) {
-        char ch = characters_without_null_termination()[i];
-        if (ch == '\n') {
-            split_view = true;
-            if (last_ch_was_cr) {
+
+    for (size_t i = 0; i < string.length(); ++i) {
+        char ch = characters[i];
+        bool split_view = false;
+
+        switch (ch) {
+        case '\n':
+            if (last_ch_was_cr)
                 substart = i + 1;
-                split_view = false;
-            }
-        }
-        if (ch == '\r') {
+            else
+                split_view = true;
+
+            last_ch_was_cr = false;
+            break;
+
+        case '\r':
             split_view = true;
             last_ch_was_cr = true;
-        } else {
+            break;
+
+        default:
             last_ch_was_cr = false;
+            break;
         }
+
         if (split_view) {
-            size_t sublen = i - substart;
-            v.append(substring_view(substart, sublen));
+            callback(string.substring_view(substart, i - substart));
             substart = i + 1;
         }
-        split_view = false;
     }
-    size_t taillen = length() - substart;
-    if (taillen != 0)
-        v.append(substring_view(substart, taillen));
-    return v;
+
+    if (size_t taillen = string.length() - substart; taillen != 0)
+        callback(string.substring_view(substart, taillen));
+}
+
+Vector<StringView> StringView::lines(bool consider_cr) const
```

Comment:
While you're here (and if you want), it would be nice to convert `bool consider_cr` to an `enum`.

Reply:
done